xbmc/network/WakeOnAccess.cpp

```diff
@@ -68,6 +73,117 @@ static unsigned long HostToIP(const std::string& host)
   return inet_addr(ip.c_str());
 }
 
+#define LOCALIZED(id) g_localizeStrings.Get(id)
+
+static void ShowDiscoveryMessage(const char* function, const char* server_name, bool new_entry)
+{
+  std::string message, heading = LOCALIZED(13033);
+
+  if (new_entry)
+  {
+    CLog::Log(LOGDEBUG, "%s - Create new entry for host '%s'", function, server_name);
+    message = StringUtils::Format(LOCALIZED(13035).c_str(), server_name);
+  }
+  else
+  {
+    CLog::Log(LOGDEBUG, "%s - Update existing entry for host '%s'", function, server_name);
+    message = StringUtils::Format(LOCALIZED(13034).c_str(), server_name);
+  }
+  CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, heading, message, 4000, true, 3000);
+}
+
+struct UPnPServer
+{
+  UPnPServer()
+  {
+    nextWake = CDateTime::GetCurrentDateTime();
+  }
+  bool operator == (const UPnPServer& server) const { return server.uuid == uuid; }
+  bool operator == (const std::string& server_uuid) const { return server_uuid == uuid; }
+  std::string name;
+  std::string uuid;
+  std::string mac;
+  CDateTime nextWake;
+};
+
+typedef std::vector<UPnPServer> upnp_server_list_t;
+
+static upnp_server_list_t upnp_servers;
+
+static UPnPServer* LookupUPnPServer(const std::string& uuid)
+{
+  upnp_server_list_t::iterator server_it = find(upnp_servers.begin(), upnp_servers.end(), uuid);
+
+  return server_it != upnp_servers.end() ? &(*server_it) : 0;
+}
+
+static void AddOrUpdateUPnPServer(const UPnPServer& server)
+{
+  upnp_server_list_t::iterator server_it = find(upnp_servers.begin(), upnp_servers.end(), server);
+
+  bool add_new_entry = server_it == upnp_servers.end();
+
+  if (add_new_entry)
+    upnp_servers.push_back(server); // add server
+  else
+    *server_it = server; // update existing server
+
+  ShowDiscoveryMessage(__FUNCTION__, server.name.c_str(), add_new_entry);
+}
+
+static CDateTime upnpInitReady;
+
+static void AddMatchingUPnPServers(const std::string& host, const std::string& mac, const CDateTimeSpan& wakeupDelay)
+{
+#ifdef HAS_UPNP
+  while (CDateTime::GetCurrentDateTime() < upnpInitReady)
+    Sleep(1000);
+
+  PLT_SyncMediaBrowser* browser = UPNP::CUPnP::GetInstance()->m_MediaBrowser;
+
+  if (browser)
+  {
+    UPnPServer server;
+    server.nextWake += wakeupDelay;
+
+    for (NPT_List<PLT_DeviceDataReference>::Iterator device = browser->GetMediaServers().GetFirstItem(); device; ++device)
+    {
+      if (host == (const char*) (*device)->GetURLBase().GetHost())
+      {
+        server.name = (*device)->GetFriendlyName();
+        server.uuid = (*device)->GetUUID();
+        server.mac = mac;
+
+        AddOrUpdateUPnPServer(server);
```

Comment:
Is it possible to get multiple matches? If not you can break out of the for-loop once you found a match.